src/components/routes/Booking/BookingOptions/BookingOptionsCrypto/BookingOptionsCrypto.tsx

```diff
@@ -51,85 +53,92 @@ const BookingOptionsCrypto = ({ booking, currency, history }: Props) => (
-                  );
-                }
-                return (
-                  <div>
-                    {accounts && (
-                      <div className="booking-options-disclaimer">
-                        <p>
-                          <span>Note:&nbsp;</span>
-                          Wallet Address is automatically updated to the current wallet selected in your Metamask
-                          browser application. Make sure the correct wallet is selected and matches the wallet address
-                          below.
-                        </p>
+                    <div>
+                      {accounts && (
+                        <div className="booking-options-disclaimer">
+                          <p>
+                            <span>Note:&nbsp;</span>
+                            Wallet Address is automatically updated to the current wallet selected in your Metamask
+                            browser application. Make sure the correct wallet is selected and matches the wallet address
+                            below.
+                          </p>
+                        </div>
+                      )}
+                      <div className="crypto-button-container">
+                        <Button
+                          className="back-button"
+                          background="light"
+                          onClick={() => history.push(`/listings/${booking.listingId}`)}
+                        >
+                          Back
+                        </Button>
+                        <SelectPaymentButton
+                          booking={booking}
+                          currency={outputCurrency}
+                          disabled={hasInsufficientFunds}
+                          onSuccess={() => history.push(`/bookings/${booking.id}/payment`)}
+                        />
                       </div>
-                    )}
-                    <div className="crypto-button-container">
-                      <Button
-                        className="back-button"
-                        background="light"
-                        onClick={() => history.push(`/listings/${booking.listingId}`)}
-                      >
-                        Back
-                      </Button>
-                      <SelectPaymentButton
-                        booking={booking}
-                        currency={currency}
-                        disabled={hasInsufficientFunds}
-                        onSuccess={() => history.push(`/bookings/${booking.id}/payment`)}
-                      />
                     </div>
-                  </div>
-                );
-              }}
-            </AppConsumer>
-          </BookingOptionsCryptoContainer>
+                  );
+                }}
+              </AppConsumer>
+            </BookingOptionsCryptoContainer>
+          );
+        };
+        return (
+          <Async promise={renderContents()} then={contents => contents} />
```

Comment:
im not sure i understand this Async component usage.

You are saying we need to wait on:
```
  const availableFunds = await getAvailableAmount(accounts, currency); 
```
Is this needed before the component loads?
If so, could this have happened in ``didComponentMount``?

Is `accounts` or `currency` being changed from a higher component? if so, could the prop just be changed and this component be re-rendered?

Alternatively, would something like HOC take care of the currency concerns and pass down to this rendered component? https://reactjs.org/docs/higher-order-components.html

Reply:
> Is this needed before the component loads?

It's needed before I can display the user's available DAI. The case is different from BEE/ETH, because those values are available as soon as Web3Provider provides `accounts`, but that pattern won't scale going forward (we'd need to query balances on every ERC-20 coin we support prior to passing `accounts` to the consumer)

> If so, could this have happened in `didComponentMount`?

Only with a larger refactor. Retrieving the balance depends on the address from `accounts`, which will not be available until after render.

> Is `accounts` or `currency` being changed from a higher component? if so, could the prop just be changed and this component be re-rendered?

`accounts` comes from the usage of Web3Provider in this component; `currency` comes from props.

> Alternatively, would something like HOC take care of the currency concerns and pass down to this rendered component?

Applying a HOC to the whole component would run into the same issue as `componentDidMount`: Don't have `accounts` yet.

---

One option here would be to change the usage of Async here to retain more of the original structure:

    <Async promise={getAvailableAmount(accounts, currency)} then={availableFunds => {
      const total = fromBee ? fromBee(quote.guestTotalAmount) : quote.guestTotalAmount;
      // ...etc...
    }} />

Or, I could just do that larger refactor and split this out into parts so we can introduce a HOC/`componentDidMount` at the appropriate level.

Reply:
the `<Async>` component is such a powerful package...but IT'S SO SIMPLE.  Why didn't we think of this😅

Reply:
i think ``<Async promise={getAvailableAmount(accounts, currency)}`` is more understandable what you are doing than ``<Async promise={renderContents()} then={contents``